### PR TITLE
Add XML-RPC interaction with staging builders

### PR DESCRIPTION
**Pull request by simonepelosi**
Updated old information to interact with builder locally.
Added documentation to interact with XML-RPC and our builder on our staging environments.
Added proxy configuration with XML-RPC.

### CODE REVIEW

File: how-to/develop-with-buildd.rst [hunk 1]
```diff
@@ -169,20 +169,43 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
+First, we need a base image to use for the builds. Usually, this is pulled as part of a build, but if we don't have Launchpad involved, we need to set this up manually.
+To download the image we need to set up the ``ubuntu-archive-tools``.
 
 .. code-block:: sh
 
     $ git clone https://git.launchpad.net/ubuntu-archive-tools
     $ sudo apt install python3-launchpadlib python3-ubuntutools
+
+Now we can download the image, note that there are two types of images: ``chroot`` and ``lxd`` images.
+``chroot`` backend is only used for ``binarypackagebuilds`` and ``sourcepackagerecipebuilds`` while most other build types use a LXD image.
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:181]: When you write only for x and y, then all other build types would use the LXD image, not only "most". What is correct?

File: how-to/develop-with-buildd.rst [hunk 2]
```diff
@@ -169,20 +169,43 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
+
+
+To download ``lxd`` image proceed as follows:
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:184]:
```suggestion
To download the ``lxd`` image proceed as follows:
```

File: how-to/develop-with-buildd.rst [hunk 3]
```diff
@@ -169,20 +169,43 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
-First, we need a base image to use for the builds. Usually this is pulled as part of a build, but if we don't have Launchpad involved, we need to set this up manually.
+First, we need a base image to use for the builds. Usually, this is pulled as part of a build, but if we don't have Launchpad involved, we need to set this up manually.
+To download the image we need to set up the ``ubuntu-archive-tools``.
 
 .. code-block:: sh
 
     $ git clone https://git.launchpad.net/ubuntu-archive-tools
     $ sudo apt install python3-launchpadlib python3-ubuntutools
+
+Now we can download the image, note that there are two types of images: ``chroot`` and ``lxd`` images.
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:180]:
```suggestion
Now we can download the image. Please note that there are two types of images: ``chroot`` and ``lxd`` images.
```

File: how-to/develop-with-buildd.rst [hunk 4]
```diff
@@ -268,3 +292,88 @@ The buildd that you have just installed needs registering with Launchpad so that
...
+First of all, we should be able to ssh into our bastion and ssh into the ``launchpad-buildd-manager`` unit since it's the one that has the firewall rules to talk with builders.
+
+Then we should follow the same procedure to get the correct ``sha1sum`` of a backend image.
+
+.. code-block:: sh
+
+    $ ./manage-chroot -s bionic -a amd64 -i lxd get 
+    $ sha1sum livecd.ubuntu-base.lxd.tar.gz
+
+The command above will show us also the link from which the image is retrieved, let's save it since we will use it in our ``ensurepresent`` function.
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:309]: That makes no sense for someone who has not worked with this before. Do not say what you want to do, but why and what is the bigger picture of it.

File: how-to/develop-with-buildd.rst [hunk 5]
```diff
@@ -268,3 +292,88 @@ The buildd that you have just installed needs registering with Launchpad so that
...
+At this point we should select the builder that we want to interact with, we can navigate ``https://qastaging.launchpad.net/builders/qastaging-bos03-amd64-001`` and get the builder location.
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:310]:
```suggestion
At this point we should select the builder that we want to interact with. We can navigate to ``https://qastaging.launchpad.net/builders/qastaging-bos03-amd64-001`` and get the builder location.
```

File: how-to/develop-with-buildd.rst [hunk 6]
```diff
@@ -268,3 +292,88 @@ The buildd that you have just installed needs registering with Launchpad so that
...
+  proxy.status()
+
+  # Inject the backend image we retrieved before.
+  proxy.ensurepresent("<sha1sum from the previous step>", "<url from the previous step>", "admin", "admin")
+
+  # Start the build.
+  proxy.build('1-3', 'snap', '<sha1sum from the previous step>', {}, {'name': 'test-build', 'image_type': 'lxd', 'series': 'bionic', 'arch_tag': 'amd64', 'git_repository': 'https://github.com/tomwardill/test-docker-repo.git', 'archives': ['deb http://archive.ubuntu.com/ubuntu bionic main restricted', 'deb http://archive.ubuntu.com/ubuntu bionic-updates main restricted', 'deb http://archive.ubuntu.com/ubuntu bionic universe']})
+
+  # After that we can clean the builder.
+  proxy.clean()
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:326]: Directly? Or would we wait for the build to be finished?

File: how-to/develop-with-buildd.rst (hunk 3 above)
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst]: Consider these 2 punctuation changes (full-stop and colon) and say "**all** other build types" to be more explicit.

> Now we can download the image. Note that there are two types of images: ``chroot`` and ``lxd`` images: ``chroot`` backend is only used for ``binarypackagebuilds`` and ``sourcepackagerecipebuilds`` while **all** other build types use a LXD image.

File: how-to/develop-with-buildd.rst (hunk 2 above)
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst]: > To download **an** ``lxd``...

Likewise, below

> To download **a** ``chroot``...

File: how-to/develop-with-buildd.rst [hunk 7]
```diff
@@ -169,20 +169,43 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
 
 .. code-block:: sh
 
     $ git clone https://git.launchpad.net/ubuntu-archive-tools
     $ sudo apt install python3-launchpadlib python3-ubuntutools
+
+Now we can download the image, note that there are two types of images: ``chroot`` and ``lxd`` images.
+``chroot`` backend is only used for ``binarypackagebuilds`` and ``sourcepackagerecipebuilds`` while most other build types use a LXD image.
+
+
```
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst:183]: Remove extra new line?

**Review by jugmac00 — CHANGES_REQUESTED** [review]
I left a couple of comments. Please ping me once addressed.

File: how-to/develop-with-buildd.rst [hunk 8]
```diff
@@ -193,10 +216,11 @@ You can try running a build via the XML-RPC interface. Start a Python/IPython re
   proxy.status()
 
 Assuming that works, a sample build can be created using (relying on the OCI capabilities being merged into launchpad-buildd):
+Note that if we are using the ``lxd`` backend we should specify that in our build ``args`` adding ``"image_type": "lxd"``.
```
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst:219]: > ...we should specify that in our build ``args`` **by** adding ``"image_type": "lxd"``.

File: how-to/develop-with-buildd.rst [hunk 9]
```diff
@@ -268,3 +292,88 @@ The buildd that you have just installed needs registering with Launchpad so that
     - Non-virtualized builders are much simpler: ``launchpad-buildd`` is cleaned synchronously over XML-RPC at the end of each build, and that's it.  If you use this, then you must be careful not to run any untrusted code on the builder (since a ``chroot`` or container escape could compromise the builder), and you'll need to uncheck "Require virtualized builders" on any PPAs, live file systems, recipes, etc. that you want to be allowed to build on this builder.
 
  4. After 30 seconds or so, the status of the builder on the builders page should be 'Idle'. This page does not auto-update, so refresh!
+
+Running a build on qastaging
+----------------------------
+
+We can use XML-RPC to interact also with qastaging/staging builders. 
+First of all, we should be able to ssh into our bastion and ssh into the ``launchpad-buildd-manager`` unit since it's the one that has the firewall rules to talk with builders.
```
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst:300]: Maybe be a little bit more specific for a new comer to understand that you'd need to `sudo -iu stg-launchpad` then `in-model=qastaging juju ssh...` type of deal

File: how-to/develop-with-buildd.rst (hunk 4 above)
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst:309]: I'd put a full stop here.

> from which the image is retrieved. Let's save it

File: how-to/develop-with-buildd.rst [hunk 10]
```diff
@@ -268,3 +292,88 @@ The buildd that you have just installed needs registering with Launchpad so that
...
+  # Start the build.
+  proxy.build('1-3', 'snap', '<sha1sum from the previous step>', {}, {'name': 'test-build', 'image_type': 'lxd', 'series': 'bionic', 'arch_tag': 'amd64', 'git_repository': 'https://github.com/tomwardill/test-docker-repo.git', 'archives': ['deb http://archive.ubuntu.com/ubuntu bionic main restricted', 'deb http://archive.ubuntu.com/ubuntu bionic-updates main restricted', 'deb http://archive.ubuntu.com/ubuntu bionic universe']})
+
+  # After that we can clean the builder.
+  proxy.clean()
+
+Proxy setup
+-----------
+
+Most of the time in our staging environments we will need to set up the proxy for our builders, with the word ``proxy`` we are referring here to both builder proxy and fetch service.
```
Comment (ineesalmeida) [inline how-to/develop-with-buildd.rst:331]: >  with the word ``proxy`` we are referring here to both builder proxy and fetch service.

I'd rather say

>  with the word ``proxy`` we are referring here to either builder proxy and fetch service. The instructions below are for setting up the builder proxy specifically.

**Review by ineesalmeida — COMMENTED** [review]
Looking good, thanks for adding all of these details to our documentation. Left a few comments to be addressed

File: how-to/develop-with-buildd.rst [hunk 11]
```diff
@@ -169,20 +169,42 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
+First, we need a base image to use for the builds. Usually, this is pulled as part of a build, but if we don't have Launchpad involved, we need to set this up manually.
+To download the image we need to set up the ``ubuntu-archive-tools``.
 
 .. code-block:: sh
 
     $ git clone https://git.launchpad.net/ubuntu-archive-tools
     $ sudo apt install python3-launchpadlib python3-ubuntutools
+
+Now we can download the image. Please note that there are two types of images: ``chroot`` and ``lxd`` images.
+``chroot`` backend is only used for ``binarypackagebuilds`` and ``sourcepackagerecipebuilds`` while all the other build types use an LXD image.
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:181]:
```suggestion
``chroot`` backend is only used for ``binarypackagebuilds`` and ``sourcepackagerecipebuilds`` while all the other build types use an ``lxd`` image.
```

File: how-to/develop-with-buildd.rst [hunk 12]
```diff
@@ -169,20 +169,42 @@ You probably want the next section (:ref:`Configuring Launchpad <configuring-lau
...
+
+To download the an ``lxd`` image proceed as follows:
```
Comment (jugmac00) [inline how-to/develop-with-buildd.rst:183]:
```suggestion
To download the ``lxd`` image proceed as follows:
```

File: how-to/develop-with-buildd.rst (hunk 6 above)
Comment (jugmac00) [reply how-to/develop-with-buildd.rst:327]: This comment has not been addressed yet.

**Review by jugmac00 — APPROVED** [review]
Looks good now!

A few remaining issues to be resolved.

This is pretty hard stuff - ideally this would be tested by someone else on a real case.

File: how-to/develop-with-buildd.rst (hunk 6 above)
Comment (simonepelosi) [reply how-to/develop-with-buildd.rst:327]: Oh sorry, I missed that. We should wait that the build finishes and then we can clean the build. I should update the comment here :)

**Pull request merged**